src/x86/runtime.rs

```diff
@@ -0,0 +1,121 @@
+    ("sse4.1") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::sse4_1{}) };
+    ("sse4.2") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::sse4_2{}) };
+    ("avx") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::avx{}) };
+    ("avx2") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::avx2{}) };
+    ("fma") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::fma{}) };
+    ("bmi") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::bmi{}) };
+    ("bmi2") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::bmi2{}) };
+    ("abm") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::abm{}) };
+    ("lzcnt") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::abm{}) };
+    ("tbm") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::tbm{}) };
+    ("popcnt") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::popcnt{}) };
+}
+
+/// X86 CPU Feature
+#[doc(hidden)]  // Please do not use this, it is an implementation detail subject to change.
+#[allow(non_camel_case_types)]
+#[repr(u8)]
+pub enum __Feature {
+    /// SSE (Streaming SIMD Extensions)
+    sse = 0,
+    /// SSE2 (Streaming SIMD Extensions 2)
+    sse2 = 1,
+    /// SSE3 (Streaming SIMD Extensions 3)
+    sse3 = 2,
+    /// SSSE3 (Supplemental Streaming SIMD Extensions 3)
+    ssse3 = 3,
+    /// SSE4.1 (Streaming SIMD Extensions 4.1)
+    sse4_1 = 4,
+    /// SSE4.2 (Streaming SIMD Extensions 4.2)
+    sse4_2 = 5,
+    /// AVX (Advanced Vector Extensions)
+    avx = 6,
+    /// AVX2 (Advanced Vector Extensions 2)
+    avx2 = 7,
+    /// FMA (Fused Multiply Add)
+    fma = 8,
+    /// BMI1 (Bit Manipulation Instructions 1)
+    bmi1 = 9,
+    /// BMI1 (Bit Manipulation Instructions 2)
+    bmi2 = 10,
+    /// ABM (Advanced Bit Manipulation) on AMD / LZCNT (Leading Zero Count) on Intel
+    abm = 11,
+    /// TBM (Trailing Bit Manipulation)
+    tbm = 12,
+    /// POPCNT (Population Count)
+    popcnt = 13,
+
+    #[doc(hidden)]
+    __NonExhaustive
+}
+
+fn set_bit(x: usize, bit: u32) -> usize {
+    debug_assert!(32 > bit);
+    x | 1 << bit
+}
+
+fn test_bit(x: usize, bit: u32) -> bool {
+    debug_assert!(32 > bit);
+    x & (1 << bit) != 0
+}
+
+fn inv_test_bit(v: usize, idx: u32) -> bool {
+    debug_assert!(32 > idx);
+    ((v >> idx) & 1) != 0
+}
+
+fn detect_features() -> usize {
+    let ebx;
+    let ecx;
+    let edx;
+
+    unsafe {
+        asm!("cpuid" : "={ecx}"(ecx), "={edx}"(edx)
+             : "{eax}"(0x00000001u32), "{ecx}"(0 as u32) : :);
+        asm!("cpuid" : "={ebx}"(ebx) : "{eax}"(0x00000007u32), "{ecx}"(0 as u32) : :);
+    }
+
+    let mut value: usize = 0;
+
+    // CPUID call with EAX=7, ECX=0 => Extended Features in EBX and ECX (unneeded):
```

Comment:
Can this link to a document online explaining what the bits are?

Reply:
Done.